cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,309 @@
+Furthermore, Cypher has good semantics for comparability and orderability for integers, floats, and strings, within each of the types.
+However working with values of different types can be difficult:
+
+  * Comparability between values of different types is often undefined. This stops query execution instead of allowing graceful recovery. This problem is particularly pronounced when it occurs as part of the evaluation of predicates (in `WHERE`).
+  * `ORDER BY` will often fail with an error if the values passed to it have different types.
+
+2. The underlying conceptual model is complex and sometimes inconsistent. This leads to an unclear relationship between comparison operators, equality, grouping, and `ORDER BY`:
+
+  * Comparability and orderability are not aligned with each other consistently, as some types may be ordered but not compared.
+  * There are various inconsistencies around equality (and equivalence) semantics as exposed by `IN`, `=`, `DISTINCT`, and grouping.
+  The difference between equality and equivalence in Cypher today is small and subtle, and limited to testing two instances of the value `null` to each other.
+
+  ** In equality, `null = null` is `null`.
+  ** In equivalence, used by `DISTINCT` and when grouping values, two `null` values are always treated as being the same value.
+  ** However, equality treats `null` values differently if they are an element of a list or a map value.
+  ** Similar rules apply for `NaN`.
+
+Furthermore, the key concepts comparability, orderability, equality, and equivalence have never been defined properly.  Therefore another motivation for this CIP is to unambiguously and precisely define these 4 concepts.
+
+== Background
+
+The reader should be mindful of the Cypher type system when reading this CIP. At the time of this proposal, the latest CIP regarding the Cypher type system is: https://github.com/opencypher/openCypher/blob/master/cip/CIP2015-09-16-public-type-system-type-annotation.adoc[`CIP2015-09-16-public-type-system-type-annotation`].
+
+== Proposal
+
+We propose to redefine <<comparability-def>> and <<equality-def>>, as well as <<orderability-def>> and <<equivalence-def>> as detailed in this section.
+
+[[concepts]]
+=== Concepts
+
+Cypher today features four distinct concepts related to equality and ordering:
+
+[horizontal]
+[[comparability-con,comparability (concept)]]*Comparability*:: <<comparability-def,Comparability>> revolves around the inequality operators (`>`, `<`, `>=`, `\<=`), and defines the semantics they need to properly determine how two values compare to each other.
+[[equality-con,equality (concept)]]*Equality*:: <<equality-def,Equality>> revolves around the equality operators ('=', '<>') as well as the list membership operator (`IN`) and value joins, and defines the semantics the needed to properly determine if two values are the same.
```

Comment:
I propose rephrase `Equality ... the equality operators, the list membership operator, and value joins`; replace `as well as` with a comma.